### PR TITLE
Multi Precision Lamb Update operator

### PR DESCRIPTION
**Pull request by access2rohit**
## Description ##
adding to new operators:

  - mp_lamb_update_phase1
  - mp_lamb_update_phase1
    Link to paper: https://arxiv.org/pdf/1904.00962.pdf

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- [x] Code is well-documented: 
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

## Testing ##
```
$ MXNET_TEST_COUNT=1 nosetests --logging-level=DEBUG --verbose -s tests/python/unittest/test_optimizer.py:test_lamb
/home/ubuntu/anaconda3/lib/python3.6/site-packages/h5py/__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters
[INFO] Setting module np/mx/python random seeds, use MXNET_MODULE_SEED=1104492960 to reproduce.
test_optimizer.test_lamb ... [DEBUG] Setting test np/mx/python random seeds, use MXNET_TEST_SEED=1864085207 to reproduce.
ok

----------------------------------------------------------------------
Ran 1 test in 9.091s

OK
```

### CODE REVIEW

File: python/mxnet/optimizer/optimizer.py [hunk 1]
```diff
@@ -1244,6 +1245,89 @@ def update(self, index, weight, grad, state):
             kwargs = {}
             sgd_update(weight, grad, out=weight, lr=lr, wd=wd, **kwargs)
 
+
+@register
+class LAMB(Optimizer):
+    """LAMB Optimizer.
+    """
+    def __init__(self, learning_rate=0.001, beta1=0.9, beta2=0.999, epsilon=1e-6,
+                 lower_bound=None, upper_bound=None, bias_correction=False, **kwargs):
```
Comment (eric-haibin-lin) [inline python/mxnet/optimizer/optimizer.py]: set default bias_correct=True since that's the recommended setting

**Comment by access2rohit** [comment]
@mxnet-label-bot add [pr-awaiting-review]

**Comment by access2rohit** [comment]
@eric-haibin-lin PR is ready for review

File: python/mxnet/optimizer/optimizer.py [hunk 2]
```diff
@@ -1262,34 +1263,73 @@ def __init__(self, learning_rate=0.001, beta1=0.9, beta2=0.999, epsilon=1e-6,
...
     def create_state(self, index, weight):
         stype = weight.stype
-        dtype = weight.dtype
-        return (zeros(weight.shape, weight.context, dtype=dtype, stype=stype),
-                zeros(weight.shape, weight.context, dtype=dtype, stype=stype))
+        return (zeros(weight.shape, weight.context, dtype=numpy.float32, stype=stype),
+                zeros(weight.shape, weight.context, dtype=numpy.float32, stype=stype))
+
+    def _update_impl(self, indices, weights, grads, states, multi_precision=False):
+        aggregate = True
```
Comment (eric-haibin-lin) [inline python/mxnet/optimizer/optimizer.py]: Not sure why you added the logic for aggregation in the code - that's used for multi-tensor updaters, which is not included in this PR. In your optimizer the indices will always be integer, right? I suggest we remove these unrelated logic in this PR

Comment (access2rohit) [reply python/mxnet/optimizer/optimizer.py]: Done

File: src/operator/optimizer_op-inl.h [hunk 3]
```diff
@@ -1749,6 +1749,164 @@ inline void LambUpdatePhaseTwo(const nnvm::NodeAttrs& attrs,
...
+    Tensor<xpu, 2, DType> grad = inputs[1].FlatTo2D<xpu, DType>(s);
+    Tensor<xpu, 2, float> mean = inputs[2].FlatTo2D<xpu, float>(s);
+    Tensor<xpu, 2, float> var = inputs[3].FlatTo2D<xpu, float>(s);
+    Tensor<xpu, 2, float> weight32 = inputs[4].FlatTo2D<xpu, float>(s);
+    Tensor<xpu, 2, float> out = outputs[0].FlatTo2D<xpu, float>(s);
+
+  Kernel<MPLambUpdatePhaseOneKernel, xpu>::Launch(s, weight.shape_.Size(),
+    out.dptr_, mean.dptr_, var.dptr_, weight.dptr_, grad.dptr_, weight32.dptr_,
+    param.clip_gradient, param.rescale_grad, beta1_t, param.beta1, beta2_t, param.beta2,
+    param.wd, param.epsilon, static_cast<int>(param.t), param.bias_correction, req[0]);
```
Comment (access2rohit) [inline src/operator/optimizer_op-inl.h]: will remove static_cast since `t` in `LambUpdatePhaseOneParam` is already changed to `int` in PR: https://github.com/apache/incubator-mxnet/pull/16903/files

File: tests/python/unittest/test_optimizer.py [hunk 4]
```diff
@@ -472,16 +481,21 @@ def update(self, index, weight, grad, state):
...
 
         g = mean_hat / (mx.nd.sqrt(var_hat) + self.epsilon) + wd * weight
+
         r2 = g.norm()
         # calculate lamb_trust_ratio
         r = 1. if r1 == 0. or r2 == 0. else r1 / r2
         lr *= r
         # update weight
         weight[:] -= lr * g
+        if use_multi_precision:
```
Comment (eric-haibin-lin) [inline tests/python/unittest/test_optimizer.py]: why adding this?

File: tests/python/gpu/test_operator_gpu.py [hunk 5]
```diff
@@ -422,6 +422,65 @@ def test_preloaded_multi_sgd():
...
+    mx_lrs = mx.nd.array(lrs, dtype='float32', ctx=mx.gpu(0))
+    wds = list((np.random.random(size=len(shapes)).astype('float32') + 0.1) / 1000.)
+    mx_wds = mx.nd.array(wds, dtype='float32', ctx=mx.gpu(0))
+
+    if use_master_weights:
+        weights32_arr = [arr.astype('float32') for arr in weights_arr]
+        mx_w32 = _make_ndarrays(weights32_arr)
+        mx_p_w32 = _make_ndarrays(weights32_arr)
+
+    mx.nd.multi_mp_lamb_update(
```
Comment (eric-haibin-lin) [inline tests/python/gpu/test_operator_gpu.py]: The test case does not make sense. There's no such an op `multi_mp_lamb_update ` in this PR.

Comment (access2rohit) [reply tests/python/gpu/test_operator_gpu.py]: forgot to remove old code :P
just removed

**Comment by access2rohit** [comment]
@mxnet-label-bot add [pr-awaiting-merge]
